File: backend/app/apis/v1/endpoints/organizations.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.apis.deps import get_current_user, get_db, require_scopes
from app.models.user import User
from app.schemas.organization import (
    OrganizationCreate,
    OrganizationOut,
    OrganizationUpdate,
)
from app.services import organizations as organization_service

router = APIRouter()


@router.get(
    "",
    dependencies=[Depends(require_scopes(["organization:read"]))],
    response_model=list[OrganizationOut],
)
def list_organizations(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return organization_service.list_organizations(db)


@router.post(
    "",
    dependencies=[Depends(require_scopes(["organization:write"]))],
    response_model=OrganizationOut,
    status_code=status.HTTP_201_CREATED,
)
def create_organization(
    payload: OrganizationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return organization_service.create_organization(db, payload)


@router.get(
    "/{organization_id}",
    dependencies=[Depends(require_scopes(["organization:read"]))],
    response_model=OrganizationOut,
)
def get_organization(
    organization_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return organization_service.get_organization(db, organization_id)


@router.put(
    "/{organization_id}",
    dependencies=[Depends(require_scopes(["organization:write"]))],
    response_model=OrganizationOut,
)
def update_organization(
    organization_id: int,
    payload: OrganizationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return organization_service.update_organization(db, organization_id, payload)


@router.delete(
    "/{organization_id}",
    dependencies=[Depends(require_scopes(["organization:write"]))],
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_organization(
    organization_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    organization_service.delete_organization(db, organization_id)
    return None

File: frontend/src/services/organizations.ts
import http from "../apis/http";
import type { ListParams, Page } from "../types/pagination";
import type {
  Organization,
  OrganizationCreate,
  OrganizationUpdate,
} from "../types/domain";

export async function listOrganizations(params: ListParams) {
  const { data } = await http.get<Page<Organization>>("/organizations", { params });
  return data;
}

export async function listOrganizationsAll() {
  const { data } = await http.get<Organization[]>("/organizations/all");
  return data;
}

export async function createOrganization(payload: OrganizationCreate) {
  const { data } = await http.post<Organization>("/organizations", payload);
  return data;
}

export async function updateOrganization(
  organizationId: number,
  payload: OrganizationUpdate,
) {
  const { data } = await http.put<Organization>(
    `/organizations/${organizationId}`,
    payload,
  );
  return data;
}

export async function deleteOrganization(organizationId: number) {
  await http.delete(`/organizations/${organizationId}`);
}

export async function deleteOrganizations(ids: number[]) {
  const { data } = await http.post("/organizations/batch-delete", { ids });
  return data as { deleted: number; skipped: number };
}
